### PR TITLE
Modal Component: rename to Dialog

### PR DESCRIPTION
**Pull request by davewhitley**
## Description
I have renamed everything from `modal` to `dialog` in the components package folder.

I'm not sure what all needs to be done. This PR is to get the conversation going.

See #14195 for discussion and context

## How has this been tested?
This PR has not been tested.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by aduth** [comment]
To note, this is a breaking change, and will need a backwards-compatibility story.

**Comment by afercia** [comment]
No strong objections. Worth noting design terminology is often different (and not so accurate) compared to technical terminology.

"Modal" is certainly a behavior. This kind of UI is more appropriately a "Dialog Window". Worth noting that dialogs can be [non-modal or modal](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal). Thus, to accurately describe this component, the correct term would be "modal dialog".

File: packages/components/src/modal/README.md [hunk 1]
```diff
@@ -1,30 +1,30 @@
-# Modal
+# Dialog
```
Comment (gziolo) [inline packages/components/src/modal/README.md]: Folder name should be updated as well. At the moment it's: ` packages/components/src/modal`.

**Review by gziolo — CHANGES_REQUESTED** [review]
> To note, this is a breaking change, and will need a backwards-compatibility story.

This might be very complex as this PR changes not only the name of the component but also all CSS class names.

Is there any way to keep the existing component unchanged and introduce `Dialog` component which would co-exist in harmony? @youknowriad any thoughts on this backward compatibility challenge? :)

**Review by gziolo — CHANGES_REQUESTED** [review]
> To note, this is a breaking change, and will need a backwards-compatibility story.

This might be very complex as this PR changes not only the name of the component but also all CSS class names.

Is there any way to keep the existing component unchanged and introduce `Dialog` component which would co-exist in harmony? @youknowriad any thoughts on this backward compatibility challenge? :)

**Comment by youknowriad** [comment]
I think the minimum here is:

 - Proxy `wp.components.Modal` to `wp.components.Dialog`.
 - Keep the old classNames in places (duplicate all the classNames)
 - Write a devnote about this change.

Separately as raised in #core-editor discussions too:

 - We should write a document in the handbook to clearly explain what is considered an official API and what is not.

**Comment by gziolo** [comment]
@drw158 do you still plan to work on this one or should we close this PR and the parent issue based on the feedback received?

**Comment by drw158** [comment]
It still should be done, but I don't have the technical knowledge to complete the tasks needed to get this done. I would prefer to keep the issue open until the issue is resolved. The PR can be closed until someone is able to work on it.

**Pull request closed without merging**